apps/prairielearn/src/tests/workspaceDynamicFiles.test.ts

```diff
@@ -0,0 +1,173 @@
+        'starter_code.c',
+        '// starter_code.c\n\n#include <stdio.h>\n#include "starter_code.h"\n\nint main()\n{\n    // enter your code\n}\n',
+      );
+      await checkFileContents('starter_code.h', '// starter_code.h\n');
+      await checkFileContents('subdir/other_file.txt', 'Content in subdirectory\n');
+      await checkFileContents(
+        'binary.bin',
+        '\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C\x0D\x0E\x0F',
+      );
+    });
+
+    it('creates all template files', async () => {
+      await checkFileContents('first_file.txt', 'A is WORD, while B is 35\n');
+      await checkFileContents('second_file.txt', 'The value of params.a is WORD.\n');
+      await checkFileContents('path/with/multiple/components.py', 'b = "WORD"\nprint(b)\n');
+    });
+
+    it('creates all dynamic files', async () => {
+      await checkFileContents('first_dynamic_file.py', 'a, b = b, a\nprint(a, b)\n');
+      await checkFileContents(
+        'second_dynamic_file.bin',
+        '\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09\x0A\x0B\x0C\x0D\x0E\x0F\x10\x11\x12\x13\x14\x15\x16\x17\x18\x19\x1A\x1B\x1C\x1D\x1E\x1F',
+      );
+      await checkFileContents(
+        'path/with/multiple/dynamic/components.bin',
+        '\x00\x11\x22\x33\x44\x55\x66\x77\x88\x99\xAA\xBB\xCC\xDD\xEE\xFF',
+      );
+      await checkFileContents('blank_file.txt', '');
+      await checkFileContents('reference_file.txt', 'This is included in the workspace.\n');
+      await checkFileContents('reference_to_subdir.txt', 'Test file.\n');
+    });
+
+    it('template files override static files', async () => {
+      await checkFileContents('template_and_static.txt', 'The correct answer is 35.\n');
+    });
+
+    it('dynamic files override template files', async () => {
+      await checkFileContents('template_and_dynamic.csv', 'a,b\n1,1\n2,4\n3,9');
+    });
+
+    it('creates no issues', async () => {
+      const issues = await queryRows(
+        sql.select_issues_for_variant_id,
+        { variant_id: variantId },
+        IssueSchema,
+      );
+      assert.lengthOf(issues, 0);
+    });
+  });
+
+  describe('Question with invalid dynamic files', () => {
+    createAndInitializeWorkspaceQuestion('workspaceInvalidDynamicFiles');
+
+    it('creates valid files', async () => {
+      await checkFileContents('static.txt', 'Static content\n');
+      await checkFileContents('template.txt', 'A is STRING, while B is 53\n');
+      await checkFileContents('dynamic.txt', 'This is a dynamic file.\n');
+    });
+
+    let issueErrors: { file: string; msg: string }[];
+    it('creates one issue', async () => {
+      const issues = await queryRows(
+        sql.select_issues_for_variant_id,
+        { variant_id: variantId },
+        IssueSchema,
+      );
+      assert.lengthOf(issues, 1);
+      issueErrors = issues[0].system_data?.courseErrData?.errors;
+      assert.isDefined(issueErrors);
+    });
+
+    it('issue lists all expected errors', async () => {
```

Comment:
I'd suggest combining these into a single test so you don't need the `let issueErrors` outside of them.

Reply:
Done in 29fbcb5.